### ISSUE TITLE
ZoomIn and ZoomOut buttons

### ISSUE THREAD

**Issue by verilog15** [anvaka/panzoom#78]
I'm trying to create buttons for zoom in and zoom out.
Is it possible? Will be glad to have some basic example.

Thank you!

**Comment by PyroGenesis** [comment]
Yes, you can create zoom buttons using the public API call `zoomTo()` (also mentioned in #28)

### Basic Example

- Use the quick play snippet for panzoom [here](https://github.com/anvaka/panzoom#script-attachment)
- Add two buttons for zoom-in and zoom-out in `<body>`

    `<button onclick="pz.zoomTo(20, 20, 1.25)">+</button>`
    `<button onclick="pz.zoomTo(20, 20, 0.8)">-</button>`

That's it. I've attached the example I made in case of any confusion.
[zoom_buttons.txt](https://github.com/anvaka/panzoom/files/2959756/zoom_buttons.txt)

**Comment by verilog15** [comment]
Hi @PyroGenesis, Thank for the replay. I try to use `smoothZoom` which is same as `zoomTo`, but it does not seem to work as expected. I defined the following function for zooming In:

```
        'zoomIn' (to, from) {
            if (to) {
                var transform = this.newArea.getTransform();
                var deltaX = transform.x;
                var deltaY = transform.y;
                var scale = transform.scale;
                var scaleChange = scale + 0.1;
                var offsetX = 0.1 + deltaX;
                var offsetY = 0.1 + deltaY;
                this.newArea.smoothZoom(offsetX,offsetY,scaleChange);
            }
            this.zoomIn = false;
        },

```

But the logic does not work as expected. I call the function everytime I need it to zoomIn. Don't really understand the logic behind `X`, `Y` and `Scale`. Also, `newArea` is the object which contains the panzoom

**Comment by Tickthokk** [comment]
I wound up just mocking a keydown event to get what I needed.

Here's a snippet.  I'm using `-in` and `-out` classes on my buttons.

```js
keyCode = $(this).is('.-in') ? 107 : 109;

// Simulate a keydown event
var e = document.createEvent('HTMLEvents');
e.initEvent('keydown', false, true);
e.keyCode = keyCode;
YOURPANZOOMELEMENT.parentElement.dispatchEvent(e);

// If the scale goes back to 1 (they zoom all the way back out), re-center
if (YOURPANZOOMINSTANCE.getTransform().scale <= 1)
	YOURPANZOOMINSTANCE.moveTo(0,0);
```

**Comment by PyroGenesis** [comment]
@verilog15 
Okay, I figured it out. I think you are confusing the scale as an absolute one when in fact it is relative. You also need to be careful to apply all methods to the `panzoom` object itself and not its container.

I modified your `zoom()` function so that it will work with normal as well as smooth zoom.

```js
var pz = document.querySelector('#panzoom-obj');
var controller = panzoom(pz);

function zoom(scale, isSmooth) {
    if (scale) {
        var transform = controller.getTransform();
        var deltaX = transform.x;
        var deltaY = transform.y;
        var offsetX = scale + deltaX;
        var offsetY = scale + deltaY;

        if (isSmooth) {
            controller.smoothZoom(offsetX, offsetY, scale);
        } else {
            controller.zoomTo(offsetX, offsetY, scale);
        }
    }
}
```
Then, its a simple matter of calling `zoom()` like so
```js
zoom(1.25, true);
```

Again, I've attached the [example file I created](https://github.com/anvaka/panzoom/files/2995903/panzoom_buttons.zip) if you have any confusion or if you want to test it yourself.

**Comment by verilog15** [comment]
Hi @PyroGenesis,
First of all, I really appreciate your help!
Actaully I use your first suggestion as follows:

```        zoomIn: function() {
            this.newArea.smoothZoom(20,20,1.25);
        },

        zoomOut: function() {
            var transform = this.newArea.getTransform();
            this.newArea.smoothZoom(10,10,0.5);
        },

**Comment by PyroGenesis** [comment]
The reason for the zoom-in on double click is that you were applying `panzoom()` to the container element instead of the inner zoomable one.

[Fixed jsfiddle](https://jsfiddle.net/1esrkvmz/)

**Comment by dota2Pro** [comment]
> @verilog15
> The reason for the zoom-in on double click is that you were applying `panzoom()` to the container element instead of the inner zoomable one.
> 
> You also needed to change jsfiddle behavior to load JavaScript in no-wrap mode (after body) for it to work properly
> 
> [Fixed jsfiddle](https://jsfiddle.net/1esrkvmz/)

Hey @PyroGenesis  I am following same steps as mentioned But either the library is glitching or I am doing something wrong can you please take a look at this [stackoverflow post](https://stackoverflow.com/questions/56859474/panzoom-glitches-out-when-zooming-in-with-a-keyboard) ?

**Comment by PyroGenesis** [comment]
@dota2Pro I have responded to your [issue here](https://github.com/anvaka/panzoom/issues/113#issuecomment-507955898) and your [Stack Overflow post here](https://stackoverflow.com/a/56863930/7120031). [Here is a JSFiddle in action.](https://jsfiddle.net/r7daj8L0/)

In short, use 'zoomAbs()` if you want to zoom using absolute scale and `zoomTo()` if you want to zoom using relative scale.

Maybe @anvaka can include some information from this thread to reduce the confusion for newer users.

**Comment by gabskoro** [comment]
I'm currently having problems with this one. I have set `transformOrigin: { x: 0.5, y: 0.5 }` and while zooming with the trackpad it's zooming from the middle, but with the buttons, it's zooming from the upper-left corner, any ideas?

**Comment by GoodPixe1** [comment]
I'm having the exact same problem as gabskoro, I've tried absolutely everything but I can't get the buttons to center-zoom, like when I use the mousewheel. @gabskoro Did you manage to find a solution for this?

**Comment by gabskoro** [comment]
@GoodPixe1 Not really, I kept using http://github.com/dy/pan-zoom. Managed to migrate everything except this one which was a bummer :(

**Comment by blahmonkey** [comment]
> I'm currently having problems with this one. I have set `transformOrigin: { x: 0.5, y: 0.5 }` and while zooming with the trackpad it's zooming from the middle, but with the buttons, it's zooming from the upper-left corner, any ideas?

The way I got this to work is
1. Get the coordinates of the _center_ of the div you want to zoom. You can do this using getBoundingClientRect on the element of the div
2. Call zoomAbs or smoothZoomAbs on those co-ordinates, with the new scale you want to set [this is easier to manage than relative zoom, but you could use this as well]

Part 1 can be tricky if you have a big div (bigger than screen) which you are applying panzoom. For this, I applied getBoundingClientRect on the parent of the div, then getBoundingClientRect on the div itself.

Then the center can be computed thus -- 
`      
const deltaX = window.scrollX + (parentRect.width / 2) - divRect.left;
const deltaY = window.scrollY + (parentRect.height / 2) - divRect.top;
`

I dont really have scroll but kept it in there in case the app changes in future

**Comment by anvaka** [comment]
There is an example to do this: 
* Live website: https://anvaka.github.io/panzoom/demo/buttons
* Source code: https://github.com/anvaka/panzoom/blob/3731bff2e15f2b4299405c2a59a24f30c3549a17/demo/buttons.html#L851

TL;DR version:

``` js
function handleClick(e) {
  e.preventDefault();
  let isZoomIn = e.target.id === 'zoomIn'; 

  let container = document.querySelector('#scene');

  // Find out center of the zooming area:
  let rect = container.getBBox();
  let cx = rect.x + rect.width/2;
  let cy = rect.y + rect.height/2;
  let zoomBy = isZoomIn ? 2 : 0.5;

  // and zoom in (`pz` is instance of panzoom)
  pz.smoothZoom(cx, cy, zoomBy);
  // Or if you don't need animation, usee this:
  // pz.zoomTo(cx, cy, zoomBy);
}
```

**Comment by a1rat91** [comment]
Sorry but the zoom is not centered on the screen. When zooming through the buttons, the content moves to the top left corner
